File: app/src/main/java/com/androidyug/marsrover/ui/main/PhotoAdapter.java
package com.androidyug.marsrover.ui.main;

import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.androidyug.marsrover.R;
import com.androidyug.marsrover.data.model.Camera;
import com.androidyug.marsrover.data.model.Photo;
import com.androidyug.marsrover.ui.detail.PhotoDetailActivity;
import com.androidyug.marsrover.ui.detail.PhotoDetailFragment;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import java.util.List;

/**
 * @author Nitin Anand (nitinnatural@gmail.com)
 */
public class PhotoAdapter extends RecyclerView.Adapter<PhotoAdapter.PhotoViewHolder>{

    List<Photo> listPhoto;
    Context ctx;

    PhotoAdapter(Context context, List<Photo> photos){
        this.listPhoto = photos;
        this.ctx = context;
    }



    public  static class PhotoViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        Photo photo;
        Context context;
        CardView cv;
        ImageView ivPhoto;
        TextView tvCamera;
        TextView tvCameraFullname;
        ImageButton ibReload;
        Boolean isPhotoLoaded = false;
        ProgressBar lineProgressBar;

        public PhotoViewHolder(View view) {
            super(view);

            view.setOnClickListener(this);
            cv = (CardView) view.findViewById(R.id.cv_photo);
            ivPhoto = (ImageView) view.findViewById(R.id.iv_photo);
            tvCamera = (TextView) view.findViewById(R.id.tv_camera);
            tvCameraFullname = (TextView) view.findViewById(R.id.tv_camera_fullname);
            ibReload = (ImageButton) view.findViewById(R.id.ib_reload);
            lineProgressBar = (ProgressBar) view.findViewById(R.id.line_progress);
            ibReload.setOnClickListener(this);
        }

        public void setItem(Photo p, final Context ctx){
            this.photo = p;
            this.context = ctx;

            Camera c = p.getCamera();

            loadPhotoIntoView();

            tvCameraFullname.setText("Photo taken by " + c.getFullName());
            tvCamera.setText(c.getName());
        }

        void hideProgress(){
            if (lineProgressBar!=null){
                lineProgressBar.setVisibility(View.GONE);
            }
        }

        void showProgress(){
            lineProgressBar.setVisibility(View.VISIBLE);
        }

        void loadPhotoIntoView(){
            showProgress();
            Picasso.with(context).load(photo.getImgSrc()).into(ivPhoto, new Callback() {
                @Override
                public void onSuccess() {
                    // TODO: 1/6/2016  hide the progress

                    isPhotoLoaded = true;
                    ibReload.setVisibility(View.GONE);
                    hideProgress();
//                    Toast.makeText(context, "Image loaded", Toast.LENGTH_SHORT).show();
                }

                @Override
                public void onError() {

                    hideProgress();
                    isPhotoLoaded = false;
                    ibReload.setVisibility(View.VISIBLE);
//                    Toast.makeText(context, "Image not able to load", Toast.LENGTH_SHORT).show();
                }
            });

        }

        @Override
        public void onClick(View v) {

            int id = v.getId();

            switch (id){
                case R.id.cv_photo:

                    if (!isPhotoLoaded){
                        Toast.makeText(context, "please let the photo to load", Toast.LENGTH_SHORT).show();
                        return;
                    }

                    Toast.makeText(context, "position "+ getLayoutPosition(), Toast.LENGTH_SHORT).show();
                    Intent i = new Intent(context, PhotoDetailActivity.class);
                    i.putExtra(PhotoDetailFragment.INTENT_PHOTO_DETAIL, photo);
                    context.startActivity(i);
                    break;

                case R.id.ib_reload:
                    ibReload.setVisibility(View.GONE);
                   loadPhotoIntoView();
                    break;

            }


        }
    }




    @Override
    public PhotoViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_view_row, parent, false);
        return new PhotoViewHolder(v);
    }

    @Override
    public void onBindViewHolder(PhotoViewHolder holder, int position) {

        holder.setItem(listPhoto.get(position), ctx);

    }

    @Override
    public int getItemCount() {
        return listPhoto.size();
    }

}

File: app/src/main/java/com/androidyug/marsrover/ui/main/PhotoAdapter.kt
package com.androidyug.marsrover.ui.main

import android.content.Context
import android.content.Intent
import android.support.v7.widget.CardView
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast

import com.androidyug.marsrover.R
import com.androidyug.marsrover.data.model.Camera
import com.androidyug.marsrover.data.model.Photo
import com.androidyug.marsrover.ui.detail.PhotoDetailActivity
import com.androidyug.marsrover.ui.detail.PhotoDetailFragment
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso

/**
 * @author Nitin Anand (nitinnatural@gmail.com)
 */
class PhotoAdapter internal constructor(private var ctx: Context, private var listPhoto: List<Photo>) : RecyclerView.Adapter<PhotoAdapter.PhotoViewHolder>() {


    class PhotoViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {
        private lateinit var photo: Photo
        private lateinit var context: Context
        private var cv: CardView
        private var ivPhoto: ImageView
        private var tvCamera: TextView
        private var tvCameraFullname: TextView
        private var ibReload: ImageButton
        private var isPhotoLoaded: Boolean = false
        private var lineProgressBar: ProgressBar

        init {

            view.setOnClickListener(this)
            cv = view.findViewById<View>(R.id.cv_photo) as CardView
            ivPhoto = view.findViewById<View>(R.id.iv_photo) as ImageView
            tvCamera = view.findViewById<View>(R.id.tv_camera) as TextView
            tvCameraFullname = view.findViewById<View>(R.id.tv_camera_fullname) as TextView
            ibReload = view.findViewById<View>(R.id.ib_reload) as ImageButton
            lineProgressBar = view.findViewById<View>(R.id.line_progress) as ProgressBar
            ibReload.setOnClickListener(this)
        }

        fun setItem(p: Photo, ctx: Context) {
            this.photo = p
            this.context = ctx

            val c = p.getCamera()

            loadPhotoIntoView()

            tvCameraFullname.text = "Photo taken by " + c!!.getFullName()
            tvCamera.text = c.getName()
        }

        private fun hideProgress() {
            if (lineProgressBar != null) {
                lineProgressBar.visibility = View.GONE
            }
        }

        private fun showProgress() {
            lineProgressBar.visibility = View.VISIBLE
        }

        private fun loadPhotoIntoView() {
            showProgress()
            Picasso.with(context).load(photo.getImgSrc()).into(ivPhoto, object : Callback {
                override fun onSuccess() {
                    // TODO: 1/6/2016  hide the progress

                    isPhotoLoaded = true
                    ibReload.visibility = View.GONE
                    hideProgress()
                    //                    Toast.makeText(context, "Image loaded", Toast.LENGTH_SHORT).show();
                }

                override fun onError() {

                    hideProgress()
                    isPhotoLoaded = false
                    ibReload.visibility = View.VISIBLE
                    //                    Toast.makeText(context, "Image not able to load", Toast.LENGTH_SHORT).show();
                }
            })

        }

        override fun onClick(v: View) {

            when (v.id) {
                R.id.cv_photo -> {

                    if (!isPhotoLoaded) {
                        Toast.makeText(context, "please let the photo to load", Toast.LENGTH_SHORT).show()
                        return
                    }

                    Toast.makeText(context, "position $layoutPosition", Toast.LENGTH_SHORT).show()
                    val i = Intent(context, PhotoDetailActivity::class.java)
                    i.putExtra(PhotoDetailFragment.INTENT_PHOTO_DETAIL, photo)
                    context.startActivity(i)
                }

                R.id.ib_reload -> {
                    ibReload.visibility = View.GONE
                    loadPhotoIntoView()
                }
            }


        }
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PhotoViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.list_view_row, parent, false)
        return PhotoViewHolder(v)
    }

    override fun onBindViewHolder(holder: PhotoViewHolder, position: Int) {

        holder.setItem(listPhoto[position], ctx)

    }

    override fun getItemCount(): Int {
        return listPhoto.size
    }

}
